File: python/examples/ex_vel_ctrl.py
#!/usr/bin/env python3
"""Joint velocity-control example ported from src/examples/ex_vel_ctrl.cpp."""

from __future__ import annotations

import argparse
import os
from pathlib import Path

import mj_kdl_wrapper as mjk

DEFAULT_MODEL = "third_party/menagerie/kinova_gen3/gen3.xml"
MODEL_ENV_VAR = "MJ_KDL_MODEL"

HOME_POSE = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
TARGET_POSE = [0.3, 0.5, 2.9, -2.0, 0.3, 1.2, 1.3]
KV = 2.0
MAX_VEL = 0.6
TOL = 0.01


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def build_robot(model_path: Path) -> tuple[mjk.Scene, mjk.Robot]:
    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True

    robot_spec = mjk.RobotSpec()
    robot_spec.path = str(model_path)
    spec.robots = [robot_spec]

    scene = mjk.Scene.build(spec)
    robot = mjk.Robot.from_scene(scene, "base_link", "bracelet_link")
    robot.ctrl_mode = mjk.CtrlMode.POSITION
    robot.set_joint_pos(HOME_POSE, call_forward=False)
    robot.jnt_pos_cmd = HOME_POSE[:]
    return scene, robot


def control_step(robot: mjk.Robot, dt: float, arrived: bool) -> bool:
    robot.update()
    if arrived:
        return True

    pos_cmd = robot.jnt_pos_cmd
    max_err = 0.0
    for i in range(robot.n_joints):
        err = TARGET_POSE[i] - robot.jnt_pos_msr[i]
        max_err = max(max_err, abs(err))
        vel = clamp(KV * err, -MAX_VEL, MAX_VEL)
        pos_cmd[i] += vel * dt

    if max_err < TOL:
        robot.jnt_pos_cmd = robot.jnt_pos_msr
        return True

    robot.jnt_pos_cmd = pos_cmd
    return False


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true", help="Open the custom Simulate UI.")
    args = parser.parse_args()

    model_path = Path(os.environ.get(MODEL_ENV_VAR, DEFAULT_MODEL))
    if not model_path.exists():
        raise FileNotFoundError(
            f"{model_path} does not exist. Run from a directory where that relative path exists "
            f"or set {MODEL_ENV_VAR}."
        )

    scene, robot = build_robot(model_path)
    try:
        dt = scene.spec.timestep
        sim_time = 0.0
        arrived = False

        if args.gui:
            viewer = mjk.SimulateViewer.open(robot, "ex_vel_ctrl.py")
            try:
                while viewer.is_running():
                    arrived = control_step(robot, dt, arrived)
                    if not viewer.step():
                        break
                    sim_time += dt
            finally:
                viewer.close()
        else:
            while sim_time < 5.0 and not arrived:
                arrived = control_step(robot, dt, arrived)
                robot.step()
                sim_time += dt

            max_err = max(abs(TARGET_POSE[i] - robot.jnt_pos_msr[i]) for i in range(robot.n_joints))
            status = "converged" if arrived else "timeout"
            print(f"max joint error: {max_err:.4f} rad  ({status})")
    finally:
        scene.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/examples/ex_vel_ctrl.cpp
/* ex_vel_ctrl.cpp
 * Joint velocity control: drive the Kinova GEN3 from the home pose to a target
 * pose using a proportional velocity controller, then hold.
 *
 * gen3.xml has POSITION actuators, so velocity is implemented by integrating
 * a clamped velocity command into the position setpoint each step:
 *
 *   vel[i]          = clamp(Kv * (target[i] - q[i]), -maxVel, maxVel)
 *   jnt_pos_cmd[i] += vel[i] * dt
 *
 * Requires MuJoCo Menagerie in cache.
 *
 * Usage:
 *   ex_vel_ctrl [--headless]
 *
 * With --headless runs until convergence (or 5 s timeout) and prints
 * final max joint error. */

#include "mj_kdl_wrapper/mj_kdl_wrapper.hpp"
#include "example_paths.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

static constexpr double kHomePose[7]   = { 0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708 };
static constexpr double kTargetPose[7] = { 0.3, 0.5, 2.9, -2.0, 0.3, 1.2, 1.3 };

static constexpr double kKv     = 2.0;  // proportional gain [rad/s per rad error]
static constexpr double kMaxVel = 0.6;  // max joint velocity [rad/s]
static constexpr double kTol    = 0.01; // convergence tolerance [rad]

int main(int argc, char *argv[])
{
    bool headless = false;
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == "--headless") headless = true;

    const std::string mjcf = mj_kdl_examples::menagerie_model("kinova_gen3/gen3.xml");

    mj_kdl::SceneSpec sc;
    sc.timestep   = 0.002;
    sc.add_floor  = true;
    sc.add_skybox = true;
    mj_kdl::RobotSpec r;
    r.path = mjcf.c_str();
    sc.robots.push_back(r);

    mj_kdl::Env env;
    if (!mj_kdl::init_env(&env, &sc)) {
        std::cerr << "init_env() failed\n";
        return 1;
    }

    mj_kdl::Robot robot;
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link")) {
        std::cerr << "init_robot_from_mjcf() failed\n";
        return 1;
    }

    unsigned n = static_cast<unsigned>(robot.n_joints);

    KDL::JntArray q_home(n);
    for (unsigned i = 0; i < n; ++i) q_home(i) = kHomePose[i];

    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::POSITION)) return 1;

    const double dt      = env.model->opt.timestep;
    bool         arrived = false;

    env.on_reset = [&](mj_kdl::ResetContext *) {
        mj_kdl::set_joint_pos(&robot, q_home);
        arrived = false;
    };

    mj_kdl::reset(&env);

    auto ctrl_step = [&]() {
        mj_kdl::update(&env);

        if (arrived) return;

        double max_err = 0.0;
        for (unsigned i = 0; i < n; ++i) {
            double err = kTargetPose[i] - robot.jnt_pos_msr[i];
            max_err    = std::max(max_err, std::abs(err));
            double vel = std::clamp(kKv * err, -kMaxVel, kMaxVel);
            robot.jnt_pos_cmd[i] += vel * dt;
        }

        if (max_err < kTol) {
            arrived = true;
            for (unsigned i = 0; i < n; ++i) robot.jnt_pos_cmd[i] = robot.jnt_pos_msr[i];
        }
    };

    if (headless) {
        const double timeout = 5.0;
        while (env.data->time < timeout && !arrived) {
            ctrl_step();
            mj_kdl::step(&env);
        }

        double max_err = 0.0;
        for (unsigned i = 0; i < n; ++i)
            max_err = std::max(max_err, std::abs(kTargetPose[i] - robot.jnt_pos_msr[i]));
        std::cout << "max joint error: " << std::fixed << std::setprecision(4) << max_err
                  << " rad  (" << (arrived ? "converged" : "timeout") << ")\n";
    } else {
        if (!mj_kdl::open_viewer(&env)) {
            std::cerr << "open_viewer() failed\n";
            return 1;
        }
        while (true) {
            ctrl_step();
            if (!mj_kdl::step(&env)) break;
            mj_kdl::pace_realtime(&env);
        }
    }

    mj_kdl::cleanup(&env);
    return 0;
}
